common/interfaces/card_interface/nfc_events.c

```diff
@@ -0,0 +1,221 @@
+static nfc_event_t nfc_event;
+static uint8_t card_removal_retry_counter = 0;
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTION PROTOTYPES
+ *****************************************************************************/
+
+/**
+ * @brief   Checks if card has been removed and returns status
+ * @arg     retry_count:  No. of consecutive retries for card not detected
+ *
+ * @return  true if card is not detected for consecutive retries exceeding
+ * retry_count
+ */
+static bool check_card_removed_status(uint8_t retry_count);
+
+/**
+ * @brief   Handles response from read NFC A target init api
+ *          If card is selected, card select event is set, if an error is
+ * received, other than NFC_RESP_NOT_READY, state is updated to
+ * NFC_STATE_SET_SELECT_CARD_CMD
+ */
+static void nfc_handle_card_select_resp();
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+static bool check_card_removed_status(uint8_t retry_count) {
+  uint32_t err = adafruit_diagnose_card_presence();
+  if (err != PN532_DIAGNOSE_CARD_DETECTED_RESP) {
+    card_removal_retry_counter++;
+  } else {
+    card_removal_retry_counter = 0;
+  }
+
+  if (card_removal_retry_counter > retry_count) {
+    return true;
+  }
+  return false;
+}
+
+static void nfc_handle_card_select_resp() {
+  nfc_a_tag_info nfc_tag_info;
+  uint32_t card_select_status = pn532_read_nfca_target_init_resp(&nfc_tag_info);
+  if (card_select_status == STM_SUCCESS) {
+    nfc_set_card_detect_event();
+  } else if (card_select_status != NFC_RESP_NOT_READY) {
+    nfc_state = NFC_STATE_SET_SELECT_CARD_CMD;
+  }
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+bool nfc_get_event(nfc_event_t *nfc_event_os_obj) {
+  ASSERT(nfc_event_os_obj != NULL);
+
+  if (nfc_event.event_occured == true) {
+    memcpy(nfc_event_os_obj, &nfc_event, sizeof(nfc_event_t));
+    nfc_reset_event();
+    return true;
+  }
+  return false;
+}
+
+void nfc_reset_event() {
+  memzero(&nfc_event, sizeof(nfc_event_t));
+}
+
+void nfc_set_card_detect_event() {
+  nfc_event.event_occured = true;
+  nfc_event.event_type = NFC_EVENT_CARD_DETECT;
+  nfc_state = NFC_STATE_CARD_DETECTED;
+}
+
+void nfc_set_card_removed_event() {
+  nfc_event.event_occured = true;
+  nfc_event.event_type = NFC_EVENT_CARD_REMOVED;
+  nfc_state = NFC_STATE_CARD_REMOVED;
+}
+
+void nfc_en_select_card_task() {
+  nfc_state = NFC_STATE_SET_SELECT_CARD_CMD;
+}
+
+uint32_t nfc_en_wait_for_card_removal_task() {
+  uint32_t card_presence_state = nfc_diagnose_card_presence();
```

Comment:
1. Why is there a requirement to check for card presence in this case? 
-> If we need it, then for consistency, we should also check if card is **not** present when `nfc_en_select_card_task()` is called

Reply:
As though why the check was introduced, if PN532 is not in the correct state (where a card is already selected/activated) diagnose command would return error even if a card was present in proximity, which leads to false result from the event call.
This event is only applicable when waiting for removal of an already activated card.

Such is not the case with `nfc_en_select_card_task()`, hence a check is not required there.

Reply:
Please add a comment in code explaining why this check is needed.

Reply:
added

Reply:
I think the comment was not pushed as I cannot see it in the diff

Reply:
Added in header file so resolving this comment